File: NSM/auth-service/internal/entity/secret_policy.go
package entity

import "time"

// SecretPolicy is a named, reusable bundle of path-authorization rules —
// see migrations/000027_create_secret_policies's own doc comment for why
// this mirrors Role's shape (including nullable OrganizationID: nil means
// a platform-wide policy usable by any organization whose user holds an
// assigned role, exactly like a system-wide Role) rather than a new one.
type SecretPolicy struct {
	ID             string
	OrganizationID *string
	Name           string
	Description    *string
	CreatedAt      time.Time
	UpdatedAt      time.Time
}

// PolicyEffect mirrors the secret_policy_effect Postgres enum — the
// entity-layer counterpart to policy.Effect (internal/policy), kept as a
// separate type because internal/policy deliberately has no dependency on
// internal/entity (see that package's own doc comment); repository code
// converts between the two.
type PolicyEffect string

const (
	PolicyEffectAllow PolicyEffect = "allow"
	PolicyEffectDeny  PolicyEffect = "deny"
)

// PolicyAction mirrors the secret_policy_action Postgres enum — the
// entity-layer counterpart to policy.Action, for the same reason.
type PolicyAction string

const (
	PolicyActionRead   PolicyAction = "read"
	PolicyActionCreate PolicyAction = "create"
	PolicyActionUpdate PolicyAction = "update"
	PolicyActionDelete PolicyAction = "delete"
	PolicyActionList   PolicyAction = "list"
)

// SecretPolicyRule is one secret_policy_rules row plus its joined
// secret_policy_rule_actions rows — a policy's individual (path_pattern,
// effect, actions) grant or restriction. PathPattern is stored already
// normalized (util.NormalizePolicyPathPattern) — see that function's own
// doc comment.
type SecretPolicyRule struct {
	ID          string
	PolicyID    string
	PathPattern string
	Effect      PolicyEffect
	Actions     []PolicyAction
	CreatedAt   time.Time
}

// SecretPolicyRoleAssignment is a secret_policy_role_assignments row — a
// policy granted to every member of a role, mirroring UserRole/GroupRole's
// own "the grant itself is its own row" shape.
type SecretPolicyRoleAssignment struct {
	PolicyID   string
	RoleID     string
	AssignedBy *string
	AssignedAt time.Time
}

File: NSM/frontend/src/types/secretPolicy.ts
// Mirrors auth-service/internal/dto/secret_policy.go exactly — Sprint 4
// Task 2's path-scoped secret authorization policies.

export type PolicyEffect = "allow" | "deny"
/** "rollback" (Path-Based Secret Authorization phase) is deliberately its
 * own action, not implied by "update" — a rule granting update on a path
 * does not, by itself, grant rolling that path back to an old version.
 * See permSecretsRollback/entity.PolicyActionRollback's own doc comments
 * on the backend. */
export type PolicyAction = "read" | "create" | "update" | "delete" | "list" | "rollback"

export interface PolicyRule {
  path_pattern: string
  effect?: PolicyEffect
  actions: PolicyAction[]
}

/** GET/POST/PUT /v1/secret-policies' row shape — metadata only, no rules.
 * See SecretPolicyDetail for the rule-bearing shape GET .../{policyId}
 * returns. */
export interface SecretPolicyResponse {
  id: string
  organization_id?: string | null
  name: string
  description?: string | null
  created_at: string
  updated_at: string
}

/** One secret_policy_rules row, as returned by the backend (never a bare
 * PolicyRule echoed back — this carries the assigned id too). */
export interface SecretPolicyRuleResponse {
  id: string
  path_pattern: string
  effect: PolicyEffect
  actions: PolicyAction[]
}

/** GET /v1/secret-policies/{policyId}'s body — the policy plus its full
 * rule set. */
export interface SecretPolicyDetailResponse extends SecretPolicyResponse {
  rules: SecretPolicyRuleResponse[]
}

/** POST /v1/secret-policies' body. */
export interface SecretPolicyCreateRequest {
  name: string
  description?: string
  rules: PolicyRule[]
}

/** PUT /v1/secret-policies/{policyId}'s body. rules is nil-vs-empty
 * significant on the backend (omitting it leaves existing rules
 * untouched) — this client only ever sends it when the caller actually
 * means to replace the rule set. */
export interface SecretPolicyUpdateRequest {
  name: string
  description?: string
  rules?: PolicyRule[]
}

/** POST /v1/secret-policies/{policyId}/assignments' body. */
export interface SecretPolicyAssignmentRequest {
  role_id: string
}
